compiler/rustc_parse/src/parser/item.rs

```diff
@@ -997,35 +997,41 @@ impl<'a> Parser<'a> {
     fn parse_item_foreign_mod(
         &mut self,
         attrs: &mut Vec<Attribute>,
-        unsafety: Unsafe,
+        mut unsafety: Unsafe,
     ) -> PResult<'a, ItemInfo> {
         let sp_start = self.prev_token.span;
         let abi = self.parse_abi(); // ABI?
-        match self.parse_item_list(attrs, |p| p.parse_foreign_item(ForceCollect::No)) {
-            Ok(items) => {
-                let module = ast::ForeignMod { unsafety, abi, items };
-                Ok((Ident::empty(), ItemKind::ForeignMod(module)))
+        if self.token.is_keyword(kw::Unsafe) {
+            let mut err = self.expect(&token::OpenDelim(Delimiter::Brace)).unwrap_err();
+            let (current_qual_sp, invalid_qual_sp) =
+                (self.prev_token.span.to(sp_start), self.token.uninterpolated_span());
+            if let Ok(current_qual) = self.span_to_snippet(current_qual_sp)
+                && let Ok(invalid_qual) = self.span_to_snippet(invalid_qual_sp)
+            {
+                err.span_suggestion(
+                        current_qual_sp.to(invalid_qual_sp),
+                        &format!("`{}` must come before `{}`", invalid_qual, current_qual),
+                        format!("{} {}", invalid_qual, current_qual),
+                        Applicability::MachineApplicable,
+                    );
             }
-            Err(mut err) => {
-                let current_qual_sp = self.prev_token.span;
-                let current_qual_sp = current_qual_sp.to(sp_start);
-                if let Ok(current_qual) = self.span_to_snippet(current_qual_sp) {
-                    // FIXME(davidtwco): avoid depending on the error message text
-                    if err.message[0].0.expect_str() == "expected `{`, found keyword `unsafe`" {
```

Comment:
Maybe we could just save the `TokenType` of the previous token if we just want to get rid of this FIXME without changing the logic so much...

Reply:
I think we can get rid of the entire `unsafe` keyword order related error message and just change the `Unsafe` field of the foreign mod when `unsafe` is detected and continue with the following checks. This way it will report an error in subsequent ast validity checks.

Reply:
Where is that ast validity check performed?

Reply:
https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_passes/src/ast_validation.rs#L1197-L1203

Reply:
I think we still want to emit an error here. If we don't, we might accidentally cause this to pass:

```rust
#[cfg(not_enabled)]
extern "C" unsafe {}
```

Reply:
Since foreign mods can't be declared `unsafe` in theory, I think it's superfluous to suggest putting `unsafe` before `extern` here. So we should only report ```expected `{`, found keyword `unsafe``` here.

Reply:
I agree, we shouldn't suggest anything. But we should continue to emit the "expected `{`, found `unsafe`" error.

Reply:
On the other hand, I think the `Unsafe` field of the foreign mod can be changed to `Yes(span)` so that it can later emit `extern block cannot be declared unsafe` error